src/factoid.rs

```diff
@@ -31,66 +31,98 @@ impl Brain {
 }
 
+pub struct Factoid<'a> {
+    pub key: &'a str,
+    pub pred: &'a str,
+    pub value: &'a str,
+}
+
 pub trait KnowsFactoids {
-    fn creates_factoid(&self, _: &str) -> bool;
-    fn create_factoid(&mut self, _: &str) -> Option<String>;
+    fn creates_factoid<'a, 'b: 'a>(&'a self, _: &'b str) -> Option<Factoid<'b>>;
+    fn create_factoid(&mut self, message: &str) -> Option<String> {
+        self.creates_factoid(message)
+            .and_then(|factoid| self.learn_factoid(factoid))
+    }
+    fn learn_factoid(&mut self, _: Factoid) -> Option<String>;
 }
 
 // TODO strip whitespass + punctuassion
 impl KnowsFactoids for Brain {
-    fn create_factoid(&mut self, s: &str) -> Option<String> {
-        // Drop name:
-        let s = s.to_string();
-        let name_index = s.find(":").unwrap();
-        let cleaned_string = s.clone().split_off(name_index + 1);
-
-        let iter = cleaned_string.split_whitespace();
-        let index = match iter
-            .clone()
-            .position(|pivot| pivot == "is" || pivot == "are")
-        {
-            Some(i) => i,
-            None => return None, // No verb here
-        };
-
-        let tmp: Vec<&str> = iter.collect();
-        let (k, v) = tmp.split_at(index);
-
-        let full_key = k.join(" ");
-        let full_val = v[1..].join(" ");
-
+    fn learn_factoid(&mut self, factoid: Factoid) -> Option<String> {
         let factoid_resp =
-            responder::FactoidResponder::new(full_key.clone().as_str(), full_val.clone().as_str());
+            responder::FactoidResponder::new(&factoid.key, &factoid.value);
         self.register_responder(factoid_resp);
 
-        return Some(format!(
+        Some(format!(
             "Ok, now I know that {} {} {}",
-            full_key, v[0], full_val
-        ));
+            &factoid.key, &factoid.pred, &factoid.value
+        ))
     }
 
-    fn creates_factoid(&self, s: &str) -> bool {
-        self.addressed(s) && (s.contains(" is ") || s.contains(" are "))
+    fn creates_factoid<'a, 'b: 'a>(&'a self, s: &'b str) -> Option<Factoid<'b>> {
+        if let Some(s) = self.addressed(s) {
+            let predicate_len: usize;
+            let mut key_len = 0usize;
+
+            let mut count_space = false;
+
+            for part in s.split_whitespace() {
+                if count_space {
+                    key_len += 1;
+                }
+
+                if part == "is" {
+                    predicate_len = 2;
+                    return Some(Factoid {
+                        // back up by 1 because we don't count the space between the last word of
+                        // the key and the predicate
+                        key: &s[..key_len - 1],
+                        pred: &s[key_len..][..predicate_len],
+                        value: &s[key_len + 1 + predicate_len..]
+                    });
+                } else if part == "are" {
+                    predicate_len = 3;
+                    return Some(Factoid {
```

Comment:
same deal with lifetimes here, bounding the lifetime on `Factoid`'s `&str` to be the same as input so it can be populated with views into whatever string we were handed a reference to